src/main/java/org/openrewrite/java/template/processor/Precondition.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2024 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.template.processor;
+
+import lombok.EqualsAndHashCode;
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
+
+import java.util.*;
+
+import static java.util.stream.Collectors.toCollection;
+
+@RequiredArgsConstructor
+public abstract class Precondition {
+    private static final Comparator<String> BY_USES_TYPE_FIRST = Comparator
+            .comparing((String s) -> !s.startsWith("new UsesType"))
+            .thenComparing(Comparator.naturalOrder());
+
+    abstract boolean fitsInto(Precondition p);
+
+    public Precondition prune() {
+        return this;
+    }
+
+    @Value
+    @EqualsAndHashCode(callSuper = false)
+    @RequiredArgsConstructor
+    public static class Rule extends Precondition {
+        String rule;
+
+        @Override
+        boolean fitsInto(Precondition p) {
+            if (p instanceof Rule) {
+                return ((Rule) p).rule.equals(rule);
+            } else {
+                return p.fitsInto(this);
+            }
+        }
+
+        @Override
+        public String toString() {
+            return rule;
+        }
+    }
+
+    @Value
+    @EqualsAndHashCode(callSuper = false)
+    @RequiredArgsConstructor
+    public static class Or extends Precondition {
+        Set<Precondition> preConditions;
+        int indent;
+
+        @Override
+        boolean fitsInto(Precondition p) {
+            throw new NotImplementedException();
```

Comment:
Should this be implemented?

Reply:
yes, comment is lost by renaming of file...